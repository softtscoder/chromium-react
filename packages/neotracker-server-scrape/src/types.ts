import {
  Asset,
  Client,
  ConfirmedTransaction,
  Contract,
  NEOONEDataProvider,
  nep5,
  RawAction,
  ReadClient,
} from '@neo-one/client-full';
import {
  Action as ActionModel,
  Coin as CoinModel,
  Contract as ContractModel,
  PubSub,
  QueryContext,
  Transaction as TransactionModel,
  TransactionInputOutput as TransactionInputOutputModel,
} from '@neotracker/server-db';
import BigNumber from 'bignumber.js';
import Knex from 'knex';
import { MigrationHandler } from './MigrationHandler';
import { IWriteCache } from './WriteCache';

export interface AddressRevert {
  readonly hash: string;
  readonly blockIndex: number;
  readonly transactionID?: string;
}
export interface AddressSave extends AddressRevert {
  readonly blockTime: number;
  readonly transactionHash?: string;
}
export interface AssetSave {
  readonly asset: {
    readonly type: Asset['type'] | 'NEP5';
    readonly name: string;
    readonly symbol?: string;
    readonly amount: string;
    readonly precision: number;
    readonly owner?: string;
    readonly admin?: string;
  };
  readonly transactionID: string;
  readonly transactionHash: string;
  readonly blockIndex: number;
  readonly blockTime: number;
  readonly hash?: string;
}
export interface ContractSave {
  readonly contract: Contract;
  readonly transactionID: string;
  readonly transactionHash: string;
  readonly blockIndex: number;
  readonly blockTime: number;
}
export interface SystemFeeSave {
  readonly index: number;
  readonly value: string;
}
export interface BlockData {
  readonly previous_block_id: number;
  readonly previous_block_hash: string;
  readonly validator_address_id: string;
  readonly aggregated_system_fee: string;
}
export interface Context {
  readonly db: Knex;
  readonly network: string;
  readonly makeQueryContext: () => QueryContext;
  readonly prevBlockData: BlockData | undefined;
  readonly currentHeight: number | undefined;
  readonly systemFee: IWriteCache<number, BigNumber, SystemFeeSave, number>;
  readonly nep5Contracts: { readonly [K in string]?: nep5.NEP5SmartContract };
  readonly chunkSize: number;
  readonly processedIndexPubSub: PubSub<{ readonly index: number }>;
  readonly client: ReadClient<NEOONEDataProvider>;
  readonly fullClient: Client;
  readonly migrationHandler: MigrationHandler;
  readonly blacklistNEP5Hashes: Set<string>;
  readonly repairNEP5BlockFrequency: number;
  readonly repairNEP5LatencySeconds: number;
}

export interface CoinChange {
  readonly address: string;
  readonly asset: string;
  readonly value: BigNumber;
}

export type CoinChanges = ReadonlyArray<CoinChange>;

export interface InputOutputResultAddressIDs {
  readonly [addressID: string]: {
    readonly startTransactionID: string;
    readonly startTransactionHash: string;
    readonly startTransactionIndex: number;
  };
}

export interface InputOutputResult {
  readonly assetIDs: ReadonlyArray<string>;
  readonly addressIDs: InputOutputResultAddressIDs;
  readonly coinChanges?: CoinChanges;
}

export interface TransferResult {
  readonly fromAddressID: string | undefined;
  readonly toAddressID: string | undefined;
  readonly assetID: string;
  readonly transferID: string;
  readonly coinChanges: CoinChanges;
}

export interface TransferData {
  readonly result: TransferResult;
  readonly value: BigNumber;
}

export interface ActionData<TAction> {
  readonly action: TAction;
  readonly transfer?: TransferData;
}

export interface TransactionData extends InputOutputResult {
  readonly claims: ReadonlyArray<TransactionInputOutputModel>;
  readonly inputs: ReadonlyArray<TransactionInputOutputModel>;
  readonly outputs: ReadonlyArray<{
    readonly type: string;
    readonly subtype: string;
    readonly output_transaction_id: string;
    readonly output_transaction_hash: string;
    readonly output_transaction_index: number;
    readonly output_block_id: number;
    readonly asset_id: string;
    readonly value: string;
    readonly address_id: string;
  }>;
  readonly actionDatas: ReadonlyArray<ActionData<RawAction>>;
  readonly transaction: ConfirmedTransaction;
  readonly transactionHash: string;
  readonly transactionID: string;
  readonly transactionIndex: number;
}

export interface TransactionModelData extends InputOutputResult {
  readonly claims: ReadonlyArray<TransactionInputOutputModel>;
  readonly inputs: ReadonlyArray<TransactionInputOutputModel>;
  readonly outputs: ReadonlyArray<TransactionInputOutputModel>;
  readonly actionDatas: ReadonlyArray<ActionData<ActionModel>>;
  readonly transactionModel: TransactionModel;
  readonly transactionHash: string;
  readonly transactionID: string;
  readonly transactionIndex: number;
  readonly contracts: ReadonlyArray<ContractModel>;
}

export interface ContractActionData extends InputOutputResult {
  readonly actionData: ActionData<RawAction>;
  readonly transactionHash: string;
  readonly transactionID: string;
  readonly transactionIndex: number;
}

export interface AssetData {
  readonly issued: BigNumber;
  readonly addressCount: number;
  readonly transactionCount: number;
  readonly transferCount: number;
}

export interface Assets {
  readonly [asset: string]: AssetData;
}

export interface AddressData {
  readonly transactionCount: number;
  readonly transferCount: number;
  readonly transactionID: string;
  readonly transactionHash: string;
}
export interface Addresses {
  readonly [address: string]: AddressData;
}

export interface CoinModelCreate {
  readonly type: 'create';
  readonly assetHash: string;
  readonly value: Partial<CoinModel>;
}

export function isCoinModelCreate(change: CoinModelChange): change is CoinModelCreate {
  return change.type === 'create';
}

export interface CoinModelDelete {
  readonly type: 'delete';
  readonly assetHash: string;
  readonly id: string;
}

export function isCoinModelDelete(change: CoinModelChange): change is CoinModelDelete {
  return change.type === 'delete';
}

export function isCoinModelCreateOrDelete(change: CoinModelChange): change is CoinModelCreate | CoinModelDelete {
  return isCoinModelCreate(change) || isCoinModelDelete(change);
}

export interface CoinModelPatch {
  readonly type: 'patch';
  readonly value: CoinModel;
  readonly patch: Partial<CoinModel>;
}

export function isCoinModelPatch(change: CoinModelChange): change is CoinModelPatch {
  return change.type === 'patch';
}

export type CoinModelChange = CoinModelCreate | CoinModelDelete | CoinModelPatch;
