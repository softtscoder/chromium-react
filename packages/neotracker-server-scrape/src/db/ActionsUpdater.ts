import { addressToScriptHash, RawAction } from '@neo-one/client-full';
import { createChild, serverLogger } from '@neotracker/logger';
import { Action as ActionModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { strip0x } from '../utils';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface ActionsSaveSingle {
  readonly action: RawAction;
  readonly transactionID: string;
  readonly transactionHash: string;
}
export interface ActionsSave {
  readonly actions: ReadonlyArray<ActionsSaveSingle>;
}
export interface ActionsRevert {
  readonly transactionIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class ActionsUpdater extends SameContextDBUpdater<ActionsSave, ActionsRevert> {
  public async save(context: Context, { actions }: ActionsSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_actions' });
    await Promise.all(
      _.chunk(actions, context.chunkSize).map(async (chunk) =>
        ActionModel.insertAll(
          context.db,
          context.makeQueryContext(),
          chunk.map(({ action, transactionID, transactionHash }) => ({
            id: action.globalIndex.toString(),
            type: action.type,
            block_id: action.blockIndex,
            transaction_id: transactionID,
            transaction_hash: transactionHash,
            transaction_index: action.transactionIndex,
            index: action.index,
            script_hash: strip0x(addressToScriptHash(action.address)),
            // tslint:disable no-any no-null-keyword
            message: (action as any).message === undefined ? null : encodeURIComponent((action as any).message),
            args_raw: (action as any).args === undefined ? null : JSON.stringify((action as any).args),
            // tslint:disable-enable no-any no-null-keyword
          })),
        ),
      ),
    );
  }

  public async revert(context: Context, { transactionIDs }: ActionsRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_actions' });
    await Promise.all(
      _.chunk(transactionIDs, context.chunkSize).map(async (chunk) =>
        ActionModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('transaction_id', chunk)
          .delete(),
      ),
    );
  }
}
