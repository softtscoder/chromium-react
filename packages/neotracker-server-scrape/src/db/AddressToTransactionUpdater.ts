import { createChild, serverLogger } from '@neotracker/logger';
import { AddressToTransaction as AddressToTransactionModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressToTransactionSaveSingle {
  readonly addressIDs: ReadonlyArray<string>;
  readonly transactionID: string;
}
export interface AddressToTransactionSave {
  readonly transactions: ReadonlyArray<AddressToTransactionSaveSingle>;
}
export interface AddressToTransactionRevert {
  readonly transactionIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressToTransactionUpdater extends SameContextDBUpdater<
  AddressToTransactionSave,
  AddressToTransactionRevert
> {
  public async save(context: Context, { transactions }: AddressToTransactionSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_address_to_transaction' });
    const data = _.flatMap(transactions, ({ addressIDs, transactionID }) =>
      [...new Set(addressIDs)].map((addressID) => ({
        id1: addressID,
        id2: transactionID,
      })),
    );
    await Promise.all(
      _.chunk(data, context.chunkSize).map(async (chunk) => {
        await AddressToTransactionModel.insertAll(context.db, context.makeQueryContext(), chunk);
      }),
    );
  }

  public async revert(context: Context, { transactionIDs }: AddressToTransactionRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_address_to_transaction' });
    await Promise.all(
      _.chunk(transactionIDs, context.chunkSize).map((chunk) =>
        AddressToTransactionModel.query(context.db)
          .context(context.makeQueryContext())
          .delete()
          .whereIn('id2', chunk),
      ),
    );
  }
}
