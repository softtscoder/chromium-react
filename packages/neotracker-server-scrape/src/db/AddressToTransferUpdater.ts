import { createChild, serverLogger } from '@neotracker/logger';
import { AddressToTransfer as AddressToTransferModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressToTransferSaveSingle {
  readonly addressIDs: ReadonlyArray<string>;
  readonly transferID: string;
}
export interface AddressToTransferSave {
  readonly transfers: ReadonlyArray<AddressToTransferSaveSingle>;
}
export interface AddressToTransferRevert {
  readonly transferIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressToTransferUpdater extends SameContextDBUpdater<AddressToTransferSave, AddressToTransferRevert> {
  public async save(context: Context, { transfers }: AddressToTransferSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_address_to_transfer' });
    const data = _.flatMap(transfers, ({ addressIDs, transferID }) =>
      [...new Set(addressIDs)].map((addressID) => ({
        id1: addressID,
        id2: transferID,
      })),
    );
    await Promise.all(
      _.chunk(data, context.chunkSize).map(async (chunk) => {
        await AddressToTransferModel.insertAll(context.db, context.makeQueryContext(), chunk);
      }),
    );
  }

  public async revert(context: Context, { transferIDs }: AddressToTransferRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_address_to_transfer' });
    await Promise.all(
      _.chunk(transferIDs, context.chunkSize).map(async (chunk) => {
        await AddressToTransferModel.query(context.db)
          .context(context.makeQueryContext())
          .delete()
          .whereIn('id2', chunk);
      }),
    );
  }
}
