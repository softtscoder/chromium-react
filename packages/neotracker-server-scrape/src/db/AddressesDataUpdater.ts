import { createChild, serverLogger } from '@neotracker/logger';
import { Address as AddressModel, isPostgres } from '@neotracker/server-db';
import { raw } from 'objection';
import { Addresses, Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressesDataSave {
  readonly addresses: Addresses;
  readonly blockIndex: number;
  readonly blockTime: number;
}
export interface AddressesDataRevert {
  readonly addresses: Addresses;
  readonly transactionIDs: ReadonlyArray<string>;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressesDataUpdater extends SameContextDBUpdater<AddressesDataSave, AddressesDataRevert> {
  public async save(context: Context, { addresses, blockIndex, blockTime }: AddressesDataSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_addresses_data' });
    await Promise.all(
      Object.entries(addresses).map(
        async ([address, { transactionCount, transferCount, transactionID, transactionHash }]) => {
          await AddressModel.query(context.db)
            .context(context.makeQueryContext())
            .where('id', address)
            .where('aggregate_block_id', '<', blockIndex)
            .patch({
              // tslint:disable no-any
              transaction_count: raw(`transaction_count + ${transactionCount}`) as any,
              transfer_count: raw(`transfer_count + ${transferCount}`) as any,
              // tslint:enable no-any
              last_transaction_id: transactionID,
              last_transaction_hash: transactionHash,
              last_transaction_time: blockTime,
              aggregate_block_id: blockIndex,
            });
        },
      ),
    );
  }

  public async revert(context: Context, { addresses, transactionIDs, blockIndex }: AddressesDataRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_addresses_data' });
    await Promise.all(
      Object.entries(addresses).map(async ([address, { transactionCount, transferCount }]) => {
        await AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .where('id', address)
          .where('aggregate_block_id', '>=', blockIndex)
          .patch({
            // tslint:disable no-any
            transaction_count: raw(`transaction_count - ${transactionCount}`) as any,
            transfer_count: raw(`transfer_count - ${transferCount}`) as any,
            // tslint:enable no-any
            aggregate_block_id: blockIndex - 1,
          });
      }),
    );

    const addressIDsSet = Object.keys(addresses);
    if (addressIDsSet.length > 0) {
      if (isPostgres(context.db)) {
        await context.db
          .raw(
            `
              WITH cte AS (
                SELECT a.id AS address_id, b.transaction_id, b.hash, b.block_time
                FROM address a
                LEFT OUTER JOIN (
                  SELECT address_id, transaction_id, hash, block_time
                  FROM (
                    SELECT id1 AS address_id, MAX(id2) AS transaction_id
                    FROM address_to_transaction
                    WHERE
                      id1 IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')}) AND
                      id2 NOT IN (${transactionIDs.map((id) => `'${id}'`).join(', ')})
                    GROUP BY address_to_transaction.id1
                  ) a
                  JOIN transaction b ON
                    a.transaction_id = b.id
                ) b ON
                  a.id = b.address_id
                WHERE
                  a.id IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')})
              )
              UPDATE address SET
                last_transaction_id=cte.transaction_id,
                last_transaction_hash=cte.hash,
                last_transaction_time=cte.block_time
              FROM cte
              WHERE
                id = cte.address_id
              `,
          )
          .queryContext(context.makeQueryContext());
      } else {
        await context.db
          .raw(
            `
              WITH cte AS (
                SELECT a.address_id, transaction_id, hash, block_time
                FROM (
                  SELECT id1 AS address_id, MAX(id2) AS transaction_id
                  FROM address_to_transaction
                  WHERE
                    id1 IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')}) AND
                    id2 NOT IN (${transactionIDs.map((id) => `'${id}'`).join(', ')})
                  GROUP BY id1
                ) a
                JOIN 'transaction' b ON
                  a.transaction_id = b.id
              )
              UPDATE address SET
                last_transaction_id=(select cte.transaction_id from cte where address_id = address.id),
                last_transaction_hash=(select cte.hash from cte where address_id = address.id),
                last_transaction_time=(select cte.block_time from cte where address_id = address.id)
              WHERE
                id IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')})
                `,
          )
          .queryContext(context.makeQueryContext());
      }
    }
  }
}
