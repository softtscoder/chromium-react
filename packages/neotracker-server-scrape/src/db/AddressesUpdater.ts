import { createChild, serverLogger } from '@neotracker/logger';
import { Address as AddressModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressesSave {
  readonly addresses: ReadonlyArray<Partial<AddressModel> & { readonly id: string }>;
}
export interface AddressesRevert {
  readonly addresses: ReadonlyArray<{ readonly id: string; readonly transactionID?: string }>;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressesUpdater extends SameContextDBUpdater<AddressesSave, AddressesRevert> {
  public async save(context: Context, { addresses }: AddressesSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_addresses' });
    const existingAddressIDs = await Promise.all(
      _.chunk(addresses, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map(({ id }) => id)),
      ),
    ).then((result) => _.flatMap(result).map((address) => address.id));

    const existingAddressIDsSet = new Set(existingAddressIDs);
    const toCreate = addresses.filter((address) => !existingAddressIDsSet.has(address.id));

    await Promise.all(
      _.chunk(toCreate, context.chunkSize).map(async (chunk) =>
        AddressModel.insertAll(context.db, context.makeQueryContext(), chunk),
      ),
    );
  }

  public async revert(context: Context, { addresses, blockIndex }: AddressesRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_addresses' });
    const addressModels = await Promise.all(
      _.chunk(addresses, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map(({ id }) => id)),
      ),
    ).then((result) => _.flatMap(result));

    const addressToTransactionID = _.fromPairs(addresses.map(({ id, transactionID }) => [id, transactionID]));
    const toDelete = addressModels.filter(
      (address) =>
        (address.transaction_id == undefined && address.block_id === blockIndex) ||
        (address.transaction_id != undefined && address.transaction_id === addressToTransactionID[address.id]),
    );

    await Promise.all(
      _.chunk(toDelete, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map((address) => address.id))
          .delete(),
      ),
    );
  }
}
