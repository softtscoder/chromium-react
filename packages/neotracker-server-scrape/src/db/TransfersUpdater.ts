import { addressToScriptHash, RawAction } from '@neo-one/client-full';
import { createChild, serverLogger } from '@neotracker/logger';
import { Transfer as TransferModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context, TransferData } from '../types';
import { strip0x } from '../utils';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface TransfersSaveSingle {
  readonly action: RawAction;
  readonly transferData: TransferData;
  readonly transactionID: string;
  readonly transactionHash: string;
  readonly transactionIndex: number;
}
export interface TransfersSave {
  readonly transactions: ReadonlyArray<TransfersSaveSingle>;
  readonly blockIndex: number;
  readonly blockTime: number;
}
export interface TransfersRevert {
  readonly transferIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class TransfersUpdater extends SameContextDBUpdater<TransfersSave, TransfersRevert> {
  public async save(context: Context, { transactions, blockIndex, blockTime }: TransfersSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_transfers' });
    await Promise.all(
      _.chunk(transactions, context.chunkSize).map(async (chunk) => {
        await TransferModel.insertAll(
          context.db,
          context.makeQueryContext(),
          chunk.map(
            ({ transferData: { result, value }, transactionID, transactionHash, transactionIndex, action }) => ({
              id: result.transferID,
              transaction_id: transactionID,
              transaction_hash: transactionHash,
              asset_id: strip0x(addressToScriptHash(action.address)),
              contract_id: strip0x(addressToScriptHash(action.address)),
              value: value.toString(),
              from_address_id: result.fromAddressID,
              to_address_id: result.toAddressID,
              block_id: blockIndex,
              transaction_index: transactionIndex,
              action_index: action.index,
              block_time: blockTime,
            }),
          ),
        );
      }),
    );
  }

  public async revert(context: Context, { transferIDs }: TransfersRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_transfers' });
    await Promise.all(
      _.chunk(transferIDs, context.chunkSize).map(async (chunk) => {
        await TransferModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk)
          .delete();
      }),
    );
  }
}
