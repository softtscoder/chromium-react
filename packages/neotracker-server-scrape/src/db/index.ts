export * from './BlockUpdater';
export * from './utils';
