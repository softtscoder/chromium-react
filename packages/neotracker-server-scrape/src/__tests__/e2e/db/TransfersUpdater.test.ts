import BigNumber from 'bignumber.js';
import { TransfersRevert, TransfersSave, TransfersUpdater } from '../../../db/TransfersUpdater';
import { CoinChange, TransferData } from '../../../types';
import { data } from '../../data';
import { updaterUnitTest, UpdaterUtilTestOptions } from '../../data/updaterTestUtil';

const coinChangesInit: ReadonlyArray<CoinChange> = [
  {
    address: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
    asset: 'g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049askbof4y8',
    value: new BigNumber('1'),
  },
];

const transferDataInit: TransferData = {
  result: {
    fromAddressID: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
    toAddressID: 'RIPEx5f4Zm4oCHwFWiSTaph1fPBxZacGVX',
    assetID: 'g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049askbof4y8',
    transferID: '1'.repeat(8),
    coinChanges: coinChangesInit,
  },
  value: new BigNumber('3'),
};

const coinChangesSecondary: ReadonlyArray<CoinChange> = [
  {
    address: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
    asset: 'g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049askbof4y8',
    value: new BigNumber('2'),
  },
];

const transferDataSecondary: TransferData = {
  result: {
    fromAddressID: 'APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR',
    toAddressID: 'RIPEx5f4Zm4oCHwFWiSTaph1fPBxZacGVX',
    assetID: 'g95k2ovja9toglc9e42c8eb9tghid43ew3ry804eaf1a0b4kfob049askbof4y8',
    transferID: '2'.repeat(8),
    coinChanges: coinChangesSecondary,
  },
  value: new BigNumber('4'),
};

const getInputs = (): TransfersSave => ({
  transactions: [
    {
      action: data.createRawLog({
        index: 0,
        globalIndex: new BigNumber(0),
      }),
      transferData: transferDataInit,
      transactionID: '1'.repeat(16),
      transactionHash: 'c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5',
      transactionIndex: 15,
    },
  ],
  blockIndex: 10,
  blockTime: 15,
});

const getSecondaryInputs = (): TransfersSave => ({
  transactions: [
    {
      action: data.createRawLog({
        index: 0,
        globalIndex: new BigNumber(0),
      }),
      transferData: transferDataSecondary,
      transactionID: '1'.repeat(16),
      transactionHash: 'c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5',
      transactionIndex: 15,
    },
  ],
  blockIndex: 10,
  blockTime: 15,
});

const getReversions = (): TransfersRevert => ({
  transferIDs: ['2'.repeat(8)],
});

const utilOptions: UpdaterUtilTestOptions<TransfersSave, TransfersRevert, undefined> = {
  name: `Transfers Updater`,
  createUpdater: () => new TransfersUpdater(),
  getInputs,
  getSecondaryInputs,
  getReversions,
};

updaterUnitTest(utilOptions);
