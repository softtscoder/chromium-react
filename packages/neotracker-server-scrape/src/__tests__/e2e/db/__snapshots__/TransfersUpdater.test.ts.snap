// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transfers Updater Transfers Updater: handles duplicate inserts/saves 1`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [],
  "transfer": Array [
    Object {
      "action_index": 0,
      "asset_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "block_id": 10,
      "block_time": 15,
      "contract_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "from_address_id": "APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR",
      "id": "11111111",
      "to_address_id": "RIPEx5f4Zm4oCHwFWiSTaph1fPBxZacGVX",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "1111111111111111",
      "transaction_index": 15,
      "value": "3",
    },
  ],
}
`;

exports[`Transfers Updater Transfers Updater: handles insert/save 1`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [],
  "transfer": Array [
    Object {
      "action_index": 0,
      "asset_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "block_id": 10,
      "block_time": 15,
      "contract_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "from_address_id": "APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR",
      "id": "11111111",
      "to_address_id": "RIPEx5f4Zm4oCHwFWiSTaph1fPBxZacGVX",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "1111111111111111",
      "transaction_index": 15,
      "value": "3",
    },
  ],
}
`;

exports[`Transfers Updater Transfers Updater: handles reverting inputs/saves 1`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [],
  "transfer": Array [
    Object {
      "action_index": 0,
      "asset_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "block_id": 10,
      "block_time": 15,
      "contract_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "from_address_id": "APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR",
      "id": "11111111",
      "to_address_id": "RIPEx5f4Zm4oCHwFWiSTaph1fPBxZacGVX",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "1111111111111111",
      "transaction_index": 15,
      "value": "3",
    },
  ],
}
`;

exports[`Transfers Updater Transfers Updater: handles reverting inputs/saves 2`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [],
  "transfer": Array [
    Object {
      "action_index": 0,
      "asset_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "block_id": 10,
      "block_time": 15,
      "contract_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "from_address_id": "APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR",
      "id": "11111111",
      "to_address_id": "RIPEx5f4Zm4oCHwFWiSTaph1fPBxZacGVX",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "1111111111111111",
      "transaction_index": 15,
      "value": "3",
    },
    Object {
      "action_index": 0,
      "asset_id": "ab38352559b8b203bde5fddfa0b07d8b25333333",
      "block_id": 10,
      "block_time": 15,
      "contract_id": "ab38352559b8b203bde5fddfa0b07d8b25333333",
      "from_address_id": "APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR",
      "id": "22222222",
      "to_address_id": "RIPEx5f4Zm4oCHwFWiSTaph1fPBxZacGVX",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "1111111111111111",
      "transaction_index": 15,
      "value": "4",
    },
  ],
}
`;

exports[`Transfers Updater Transfers Updater: handles reverting inputs/saves 3`] = `
Object {
  "action": Array [],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [],
  "transfer": Array [
    Object {
      "action_index": 0,
      "asset_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "block_id": 10,
      "block_time": 15,
      "contract_id": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "from_address_id": "APyEx5f4Zm4oCHwFWiSTaph1fPBxZacYVR",
      "id": "11111111",
      "to_address_id": "RIPEx5f4Zm4oCHwFWiSTaph1fPBxZacGVX",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "1111111111111111",
      "transaction_index": 15,
      "value": "3",
    },
  ],
}
`;
